test/framework/src/main/java/org/opensearch/test/AbstractBuilderTestCase.java

```diff
@@ -316,6 +323,28 @@ protected static QueryShardContext createShardContext() {
         return createShardContext(null);
     }
 
+    protected static QueryBuilderVisitor createTestVisitor() {
+        return new QueryBuilderVisitor() {
+            @Override
+            public void accept(QueryBuilder qb) throws IOException {
+                if (qb instanceof BoolQueryBuilder) {
+                    throw new IOException("Bool Query Builder Traversed");
+                } else if (qb instanceof BoostingQueryBuilder) {
+                    throw new IOException("Boosting Query Builder Traversed");
+                } else if (qb instanceof DisMaxQueryBuilder) {
+                    throw new IOException("DisMax Query Builder Traversed");
+                } else if (qb instanceof FieldMaskingSpanQueryBuilder) {
+                    throw new IOException("Field Masking Query Builder Traversed");
+                }
+            }
+
+            @Override
+            public QueryBuilderVisitor getChildVisitor(BooleanClause.Occur occur) {
+                return NO_OP_VISITOR;
+            }
+        };
+    }
```

Comment:
This visitor doesn't test the child-visiting behavior of the `visit` methods.

You're better off (IMO) doing something like:

```
protected static class CollectingQueryBuilderVisitor implements QueryVisitor {
  private final List<QueryBuilder> visited = new ArrayList<>();
  private final Map<BooleanClause.Occur, List<CollectingQueryBuilderVisitor>> childVisitors = new HashMap<>();

  @Override
  public void accept(QueryBuilder qb) {
    visited.add(qb);
  }

  @Override
  public QueryBuilderVisitor getChildVisitor(BooleanClause.Occur occur) {
    CollectingQueryBuilderVisitor childVisitor = new CollectingQueryBuilderVisitor();
    childVisitors.computeIfAbsent(occur, ArrayList<>::new).add(childVisitor);
    return childVisitor;
  }
}
```

Reply:
That way you essentially "mirror" the query structure can can verify that you got the right "shape".

Reply:
I have taken care of it now